fish-rust/src/event.rs

```diff
@@ -0,0 +1,926 @@
+const ATOMIC_BOOL_FALSE: AtomicBool = AtomicBool::new(false);
+#[allow(clippy::declare_interior_mutable_const)]
+const ATOMIC_U32_0: AtomicU32 = AtomicU32::new(0);
+
+static PENDING_SIGNALS: PendingSignals = PendingSignals {
+    counter: AtomicU32::new(0),
+    received: [ATOMIC_BOOL_FALSE; SIGNAL_COUNT],
+    last_counter: Mutex::new(0),
+};
+
+/// List of event handlers. **While this is locked to allow safely accessing/modifying the vector,
+/// note that it does NOT provide exclusive access to the [`EventHandler`] objects which are shared
+/// references (in an `Arc<T>`).**
+static EVENT_HANDLERS: Mutex<EventHandlerList> = Mutex::new(Vec::new());
+
+/// Tracks the number of registered event handlers for each signal.
+/// This is inspected by a signal handler. We assume no values in here overflow.
+static OBSERVED_SIGNALS: [AtomicU32; SIGNAL_COUNT] = [ATOMIC_U32_0; SIGNAL_COUNT];
+
+/// List of events that have been sent but have not yet been delivered because they are blocked.
+///
+/// This was part of profile_item_t accessed as parser.libdata().blocked_events and has been
+/// temporarily moved here. There was no mutex around this in the cpp code. TODO: Move it back.
+static BLOCKED_EVENTS: Mutex<Vec<Event>> = Mutex::new(Vec::new());
+
+fn inc_signal_observed(sig: usize) {
+    if let Some(sig) = OBSERVED_SIGNALS.get(sig) {
+        sig.fetch_add(1, Ordering::Relaxed);
+    }
+}
+
+fn dec_signal_observed(sig: usize) {
+    if let Some(sig) = OBSERVED_SIGNALS.get(sig) {
+        sig.fetch_sub(1, Ordering::Relaxed);
+    }
+}
+
+/// Returns whether an event listener is registered for the given signal. This is safe to call from
+/// a signal handler.
+pub fn is_signal_observed(sig: usize) -> bool {
+    // We are in a signal handler!
+    OBSERVED_SIGNALS
+        .get(sig)
+        .map_or(false, |s| s.load(Ordering::Relaxed) > 0)
+}
+
+pub fn get_desc(parser: &parser_t, evt: &Event) -> WString {
+    let s = match &evt.desc.typ {
+        EventType::Signal { signal } => format!(
+            "signal handler for {} ({})",
+            sig2wcs(*signal),
+            signal_get_desc(*signal)
+        ),
+        EventType::Variable { name } => format!("handler for variable '{name}'"),
+        EventType::ProcessExit { pid } => format!("exit handler for process {pid}"),
+        EventType::JobExit { pid, .. } => {
+            if let Some(job) = parser.job_get_from_pid(*pid) {
+                format!(
+                    "exit handler for job {}, '{}'",
+                    job.job_id().0,
+                    job.command()
+                )
+            } else {
+                format!("exit handler for job with pid {pid}")
+            }
+        }
+        EventType::CallerExit { .. } => "exit handler for command substitution caller".to_string(),
+        EventType::Generic { param } => format!("handler for generic event '{param}'"),
+        EventType::Any => unreachable!(),
+    };
+
+    WString::from_str(&s)
```

Comment:
it would be more orthodox to use `sprintf` instead of `format!`+`WString::from_str`

Reply:
Probably. But also more work if we ever want to leave WString behind. I don't think it matters either way though.

Reply:
OK. It would be nice to get rid of printf-style format specifiers eventually, maybe before we reitre WString. We could write a `wformat!`.